test/jasmine/tests/click_test.js

```diff
@@ -65,24 +66,62 @@ describe('Test click interactions:', function() {
     }
 
     describe('click events', function() {
-        var futureData;
+        var futureData, clickPassthroughs, contextPassthroughs;
 
         beforeEach(function(done) {
             Plotly.plot(gd, mockCopy.data, mockCopy.layout).then(done);
 
+            futureData = undefined;
+            clickPassthroughs = 0;
+            contextPassthroughs = 0;
+
             gd.on('plotly_click', function(data) {
                 futureData = data;
             });
+
+            gd.addEventListener('click', function() {
+                clickPassthroughs++;
+            });
+            gd.addEventListener('contextmenu', function() {
+                contextPassthroughs++;
+            });
         });
 
-        it('should not be trigged when not on data points', function() {
+        // Later we want to emit plotly events for clicking in the graph but not on data
+        // showing the axis values you clicked on. But at the moment these events
+        // pass through to event handlers attached to gd.
+        it('should not be triggered when not on data points', function() {
             click(blankPos[0], blankPos[1]);
             expect(futureData).toBe(undefined);
+            // this is a weird one - in the real case the original click never
+            // happens, it gets canceled by preventDefault in mouseup, but we
+            // add our own synthetic click.
+            // But assets/click doesn't know not to generate a click event, so
+            // here we get both. I don't see a good way to avoid this, but also
+            // there's something nice about this showing that we do indeed
+            // generate the synthetic click event.
+            // TODO: do we actually want this synthetic event, now that dragElement
+            // has `clickFn` to explicitly manage clicks too? Perhaps leave it in
+            // for now, in case either 1) users want to catch all clicks on gd, or
+            // 2) we have a component still using on('click') instead of `clickFn`
+            expect(clickPassthroughs).toBe(2);
+            expect(contextPassthroughs).toBe(0);
+        });
+
+        // Margin clicks will probably always pass through to gd, right?
+        // Any reason we should handle these?
+        it('should not be triggered when in the margin', function() {
+            click(marginPos[0], marginPos[1]);
+            expect(futureData).toBe(undefined);
+            expect(clickPassthroughs).toBe(1);
+            expect(contextPassthroughs).toBe(0);
         });
 
         it('should contain the correct fields', function() {
             click(pointPos[0], pointPos[1]);
             expect(futureData.points.length).toEqual(1);
+            expect(clickPassthroughs).toBe(2);
```

Comment:
so, in a real world scenario, this here would be `.toBe(0)`?

Reply:
i.e. the `plotly_click` eats up `gd.on('click')`?

Reply:
real world it would be `1`. `dragElement` eats up one click event but generates a new one. That was important in certain cases when we were using that synthetic event to trigger the `selection.on('click')` to trigger the `plotly_click`. We could consider removing it now, per [this TODO](https://github.com/plotly/plotly.js/pull/2241/files#diff-cb022fcf45f61114c1c73baba3164c1aR103) so it doesn't bubble up the tree after we've already generated our own events from it, but leaving it in is the most backward-compatible way to handle this.